# bioinformatics
Learning bioinformatics in Python; these three scripts illustrate the iterative improvement of a motif finding algorithm. The first iteration builds up a list of probable motifs based on a simple scoring system that tallies nucleotide mismatches at each position(GreedyMotifSearch). The second iteration introduces randomized motif building and accounts for Laplace's Principle by adding pseudocounting to the scoring system to improve the accuracy of predictions(RandomizedMotifSearch). The third and last iteration is a Gibb's Sampler that adds another layer of randomization and searches DNA strings at a more granular level by making less drastic changes at each step when compared to RandmoizedMotifSearch. Tools were used to search for the DNA binding motif of the Dormancy Survival Regulolator(DoSR) of Mycobacterium tiberculosis on a subset of the upstream regions of genes under the control of DoSR in a a 2003 study in which scientists used a DNA microarray experiment to identify the genes. Guided by the Bioinformatics Specialization on Coursera. 

In the future I would like to add a GUI with the tkinter package, edit the code to be cleaner and experiment with different outputs, also use algorithms on full data set. Possibly incorporate capabilities with the Biopython package. 
